common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+                            future
+                        }
+                    }
+            }
+
+        } else {
+            // not caching
+            future
+        }
+    }
+
+    protected def cacheUpdate(doc: W, key: Any, future: => Future[Winfo])(
+        implicit transid: TransactionId, logger: Logging, ec: ExecutionContext): Promise[Winfo] = {
+
+        val valPromise = Promise[W]
+        val promise = Promise[Winfo]
+
+        if (cacheEnabled) {
+            logger.info(this, s"caching $key")
+
+            val desiredEntry = Entry(transid, WriteInProgress, valPromise.future)
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) map {
+                // ... and see what we get back
+                case actualEntry =>
+                    if (transid == actualEntry.transid) {
+                        //
+                        // then we won the race, and now own the entry
+                        //
+                        logger.debug(this, s"Write initiated $key")
+                        listenForWriteDone(doc, key, actualEntry, future, valPromise, promise)
+
+                    } else {
+                        //
+                        // then we either lost the race, or there is an operation in progress on this key
+                        //
+                        logger.debug(this, s"Write under ${actualEntry.state.get} $key")
+                        promise completeWith future
+                    }
+            }
+
+        } else {
+            // not caching
+            promise completeWith future
+        }
+
+        promise
+    }
+
+    def getSize() = cache.size
+
+    /**
+     * Log a cache hit
+     *
+     */
+    private def makeNoteOfCacheHit(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_HIT, s"[GET] serving from cache: $key")(logger)
+    }
+
+    /**
+     * Log a cache miss
+     *
+     */
+    private def makeNoteOfCacheMiss(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_MISS, s"[GET] serving from datastore: $key")(logger)
+    }
+
+    /**
+     * We have initiated a read (in cacheLookup), now handle its completion
+     *
+     */
+    private def listenForReadDone(key: Any, entry: Entry, future: => Future[W], promise: Promise[W])(
+        implicit logger: Logging): Unit = {
+
+        future onComplete {
+            case Success(value) => entry.readDone(value, promise)
+            case Failure(t)     => readOops(key, entry, promise, t)
+        }
+    }
+
+    /**
+     * We have initiated a write, now handle its completion
+     *
+     */
+    private def listenForWriteDone(doc: W, key: Any, entry: Entry, future: => Future[Winfo], valPromise: Promise[W], promise: Promise[Winfo])(
+        implicit logger: Logging): Unit = {
```

Comment:
As noted above, this should really return `Future[Winfo]`, so error cases can be handled above and not need to thread multiple `Promise` instances through